データベース設計

sorairo580 opened the pull request:
# What
ER図の作成とREADMEの編集
↓ER図
https://gyazo.com/92f5423f749c056a3cb12adf411a6970

# Why
データベース設計のため

sorairo580 committed:
READMEの編集

koyamatakumi commented on README.md:
```diff
@@ -1,24 +1,88 @@
 
-This README would normally document whatever steps are necessary to get the
-application up and running.
+## users table
```
usersテーブルに必要なカラムと不要なカラムが混在しております。
usersテーブルはユーザーの新規登録画面で入力する内容を保存するため電話番号等のカラムは不要です。
まずは見本アプリの新規登録画面の内容を確認し、必要なカラムを洗い出してみましょう。

koyamatakumi commented on README.md:
```diff
@@ -1,24 +1,88 @@
-* Database initialization
+| Column                  | Type       | Options                        |
+|-------------------------|------------|--------------------------------|
+| item name               | string     | null: false                    |
```
```
| item name               | string     | null: false                    |
```
カラム名にスペースを含めることはできません。
「item_name」のようにアンダーバーを使用しましょう。

■補足
他のカラムにもスペースが含まれているものが確認できます。
同様に修正しましょう。

koyamatakumi commented on README.md:
```diff
@@ -1,24 +1,88 @@
+| category                | string     | null: false                    |
+| price                   | string     | null: false                    |
+| item description        | text       | null: false                    |
+| message                 | text       | null: false                    |
```
```
| message                 | text       | null: false                    |
```
不要なカラムのため消しておきましょう。

koyamatakumi commented on README.md:
```diff
@@ -1,24 +1,88 @@
+| Column                  | Type       | Options                        |
+|-------------------------|------------|--------------------------------|
+| item name               | string     | null: false                    |
+| category                | string     | null: false                    |
```
```
| category                | string     | null: false                    |
```
**各カラム名の末尾に_idと追記していただき、カラムの方はinteger型へ修正をお願いいたします。**
■補足
出品画面でプルダウンで選択する項目のカラムはActiveHashを用いての実装でございます。

ActiveHashに関しましては、最終課題補足カリキュラム【カテゴリーの選択を工夫しよう】をご参考ください。
https://master.tech-camp.in/v2/curriculums/6769

■記述例
| 〇〇_id       | integer     | null: false       |

koyamatakumi commented on README.md:
```diff
@@ -1,24 +1,88 @@
+|-------------------------|------------|--------------------------------|
+| item name               | string     | null: false                    |
+| category                | string     | null: false                    |
+| price                   | string     | null: false                    |
```
```
| price                   | string     | null: false                    |
```
integer型に修正しましょう。

koyamatakumi commented on README.md:
```diff
@@ -1,24 +1,88 @@
-* Configuration
 
-* Database creation
+## items table
```
itemsテーブルに必要なカラムが不足しております。
itemsテーブルは商品の出品画面で入力する内容を保存します。
まずは見本アプリの出品画面の内容を確認し、必要なカラムを洗い出してみましょう。

koyamatakumi commented on README.md:
```diff
@@ -1,24 +1,88 @@
-* Deployment instructions
 
-* ...
+## purchases table
```
今回必要なテーブルは計4つです。
```
・"ユーザー情報"を管理するテーブル
・"商品情報"を管理するテーブル
・"購入管理情報"（誰が何を購入したか）を管理するテーブル
・"配送先情報"を管理するテーブル
```

必要なテーブルにつきましては以下のカリキュラムに動画で解説がございます。
https://master.tech-camp.in/v2/curriculums/8125

また各テーブルに必要なカラムは見本アプリの新規登録ページや出品ページをご確認頂くと洗い出すことができます。合わせて以下のカリキュラムも参考になりますのでご覧くださいませ。
https://master.tech-camp.in/v2/curriculums/8136

koyamatakumi reviewed (commented):
sorairo580様レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認をお願いいたします。

また、READMEの修正に合わせて、ER図の修正もお願いいたします。
修正後は改めてプルリクエストのコメントにgyazoの送付をお願いいたします。

README修正後はコミットプッシュをしていただき、改めてレビュー依頼フォームの送信をお願いいたします。

ご不明ございましたらコネクトメンターにご質問頂けますと幸いです。

sorairo580 committed:
DB設計２

sorairo580 commented:
ご確認いただきありがとうございます。
修正したER図のほうも添付します。↓
https://gyazo.com/af3d944fe9a0eef2b99d55d943bd94a8

ただ、購入履歴の記述がどこまで記載すべきかわかりません。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+| first_name         | string              | null: false               |
+| last_name_kana     | string              | null: false               |
+| first_name_kana    | string              | null: false               |
+| birth_year_id      | integer             | null: false               |
```
誕生日カラムを一つにまとめましょう。
日付型を用いることで年月日を一つのカラムで管理することが可能になります。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
-* Configuration
 
-* Database creation
+## items table
```
ユーザーの外部キーカラムを追加しましょう。

■理由
実装要件において、商品テーブルには出品者が誰であるかを保存することが求められているためになります。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+| condition_id            | integer    | null: false                    |
+| cost_id                 | integer    | null: false                    |
+| area_id                 | integer    | null: false                    |
+| days_id                 | integer    | null: false                    |
```
[day]という単語はrailsの予約語です。
使用すると予期せぬエラーが出る可能性がございますので、別のカラム名に修正しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
-* Deployment instructions
 
-* ...
+## purchases-records table
```
商品の購入管理はユーザーと商品のidのみで管理が可能です。
「どのユーザーがどのアイテムと紐付いているか」を管理する役割になります。
したがって、それ以外のカラムは今回は不要です。
修正しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+
+| Column          | Type       | Options                        |
+|-----------------|------------|--------------------------------|
+| zip_cord        | integer    | null: false                    |
```
郵便番号はハイフンを必要とすることも考慮してstring型で実装しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+| city            | string     | null: false                    |
+| house_number    | string     | null: false                    |
+| room_number     | string     |                                |
+| phone_number    | integer    | null: false                    |
```
電話番号は数字の文字列として実装しておきたいためstring型で実装しましょう。

（integerは整数のため、頭に０がついた整数が存在しない以上非表示になってしまう可能性があるためです）

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+- has_many   :user_assemblys
+
+
+## delivery-destinations table
```
購入管理テーブルの外部キーを追記しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+
+### Association
+
+- belongs_to :items
```
配送先住所テーブルは購入管理テーブルと紐づいているため、商品モデルとのアソシエーションは不要です。
削除しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+### Association
+
+- belongs_to :items
+- has_one    :purchase_redords
```
配送先住所テーブルは購入管理テーブルの親子関係に着目し、アソシエーションを修正しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
+### Association
+
+- belongs_to :users
+- has_many   :user_assemblys
```
```
- has_many   :user_assemblys
```
user_assemblysテーブルは見つかりませんでした。
不要な記述であれば削除しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,24 +1,72 @@
 
-* Services (job queues, cache servers, search engines, etc.)
+- belongs_to :user
+- has_one    :delivery-destinations
```
配送先住所テーブルは購入管理テーブルと紐づいているため、ユーザーモデルとのアソシエーションは不要です。
削除しましょう。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いいたします。
修正後、再度レビュー依頼フォームにてご提出くださいませ。

最新のER図の添付をお願いいたします。

sorairo580 committed:
DB設計３

誕生日カラム、itemsテーブルのユーザー外部キーカラムの追加、days_idをdelivery_days_idへ、purchases-recordsテーブルのカラム修正及びアソシエーションの修正、郵便番号の型をstring型へ、delivery-destinationsテーブルの外部キー追加及びアソシエーションの修正

sorairo580 commented:
修正しました。
再度ご確認をお願いします。
ER図↓
https://gyazo.com/1a45f50ff91cf47091243c1756d6e044

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| first_name         | string              | null: false               |
+| last_name_kana     | string              | null: false               |
+| first_name_kana    | string              | null: false               |
+| birthday_id        | date                | null: false               |
```
◯カラム名を修正しましょう！

生年月日はid登録ではないため

birthday等の名前に修正しましょう！

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
 
-* System dependencies
+- has_many :items
+- has_many :purchase-records
```
◯モデル名を修正しましょう！
→小文字のスネークケースを使用しましょう

purchase-records

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
 
-* Services (job queues, cache servers, search engines, etc.)
+- belongs_to :user
+- has_one    :delivery-destinations
```
◯アソシエーションを修正しましょう！

購入履歴へのアソシエーションに修正しましょう

has_one :delivery-destinations

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
-* ...
+
+## purchases-records table
+
```
◎必要なカラムを見直しましょう！
→こちらのテーブルに必要なカラムは下記の通りです。


```

| ユーザー  | 外部キーカラム  | NOT NULL 制約, 外部キーオプション    |
| 商品      | 外部キーカラム  | NOT NULL 制約, 外部キーオプション    |


※各カラム名は保存する情報が分かるように命名しましょう！

```

再度見本アプリを使いながら確認していきましょう！

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
+
+### Association
+
+- has_many :delivery-destinations
```
◯アソシエーションを修正しましょう！

購入履歴に対する配送先住所は一つです

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| item_description        | text          | null: false                    |
+| condition_id            | integer       | null: false                    |
+| cost_id                 | integer       | null: false                    |
+| area_id                 | integer       | null: false                    |
```
◎都道府県に関するカラムを修正しましょう！

```
area_id | integer | null: false |
```

都道府県カラムにつきましては、商品テーブル・配送先テーブルのカラム名を統一する必要がございます。

都道府県(prefecture_id)等のカラム名にて２つのテーブルで内容を揃えていきましょう！

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| city            | string     | null: false                    |
+| house_number    | string     | null: false                    |
+| room_number     | string     |                                |
+| phone_number    | integer    | null: false                    |
```
◯以下の指摘内容の修正が確認できませんでした。

https://github.com/sorairo580/furima-39405/pull/14/files#r1320711207

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| house_number    | string     | null: false                    |
+| room_number     | string     |                                |
+| phone_number    | integer    | null: false                    |
+| item_info       | references | null: false, foreign_key: true |
```
◯カラムを修正しましょう！

必要な外部キーは購入テーブルのものです

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| item_info       | references | null: false, foreign_key: true |
+
+### Association
+
```
◎アソシエーションを修正しましょう！

belongs_toの対象は単数形

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※DB設計にて修正箇所のございます際は、お手数ですが、最新のER図のキャプチャ画像も改めて添付頂きますようお願い致します。
```

sorairo580 committed:
DB設計４

birthdayの修正及びアソシエーション削除、prifecture_idへの統一及びアソシエーション修正、purchases_recordsテーブルのカラム追加及びアソシエーション修正、小文字スネークケースへの修正、電話番号の変更及びアソシエーション修正

sorairo580 commented:
度々申し訳ありません。
DB設計４のプッシュを行いました。
ご確認をお願いします。

ER図↓
https://gyazo.com/657158341fd6390720fcdb9f17192a80

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,67 @@
 
-* System dependencies
+- has_many :items
 
```
◯購入履歴へのアソシエーションを追加しましょう

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,67 @@
 
-* Services (job queues, cache servers, search engines, etc.)
+- belongs_to :user
+- has_many   :purchases_records
```
◯アソシエーションを修正しましょう

紐づく購入履歴は一つのみです

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,67 @@
+| Column      | Type       | Options                        |
+|-------------|------------|--------------------------------|
+| user        | references | null: false, foreign_key: true |
+| item_info   | references | null: false, foreign_key: true |
```
◯カラム名を修正しましょう

| item| references | null: false, foreign_key: true |

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,67 @@
+- belongs_to :item
+
+
+## delivery_destinations table
```
◯購入テーブルの外部キーカラムを追加しましょう

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,67 @@
+
+### Association
+
+- has_one :purchase_record
```
◎アソシエーションを修正しましょう！

購入テーブルが親テーブルです

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,67 @@
+
+### Association
+
+- belongs_to :delivery_destination
```
◎アソシエーションを修正しましょう！

配送先テーブルが小テーブル

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,67 @@
+### Association
+
+- belongs_to :delivery_destination
+- belongs_to :item
```
◯アソシエーションを追加しましょう

ユーザーへのつながりが必要です

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※DB設計にて修正箇所のございます際は、お手数ですが、最新のER図のキャプチャ画像も改めて添付頂きますようお願い致します。
```

sorairo580 committed:
DB設計５

購入履歴へのアソシエーション追加、has_oneへの修正、カラム名の修正、購入テーブルの外部キーカラム追加、親テーブルと子テーブルの変更、userテーブルへのアソシエーション追加
余白の削除

sorairo580 commented:
度々申し訳ありません。
DB設計５のプッシュを行いました。
ご確認をお願いします。

ER図↓
https://gyazo.com/719283d9f281b71256cb86a21773263b

mihot117 commented on README.md:
```diff
@@ -1,24 +1,70 @@
 
-* System dependencies
+- has_many :items
+- has_many :purchase-records
```
```ruby
- has_many :purchase-records
```

カラム名はハイフンの使用ができませんので、以下のようにアンダーバーをお使いください！

```ruby
- has_many :purchase_records
```

mihot117 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| cost_id          | integer    | null: false                    |
+| prifecture_id    | integer    | null: false                    |
+| delivery_days_id | integer    | null: false                    |
+| user             | references | null: false, foreign_key: true |
```
```ruby
| delivery_days_id | integer    | null: false                    |
```

上記のみ複数名詞_idという形になってしまっています。
Active hashのカラムは外部キーカラムと同様に「単数形_id」であることが必須であるため、daysのsを抜いておきましょう！

mihot117 commented on README.md:
```diff
@@ -1,24 +1,70 @@
 
-* ...
+## purchases_records table
+
```
```ruby
## purchases_records table
```

こちらのテーブルは、`purchase`に複数形の`s`を付けている形が正しいでしょうか？

usersテーブルやitemsテーブル内のアソシエーションを確認していると、`purchases_records`になっているところと`purchase_records`になっているところの2種類が混在していますので、どちらかに統一しておきましょう。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| phone_number  | string     | null: false                    |
+| item          | references | null: false, foreign_key: true |
+
+### Association
```
```ruby
| item | references | null: false, foreign_key: true |
```

上記ですが、今回はitemだけではなくuser情報も取得したいので、itemsテーブルの外部キーカラムの代わりに`purchases_recordsテーブルの外部キーカラム`を準備しましょう！

purchases_recordsテーブルはitemとuserの情報を一括で管理しているので、同時取得が可能になります。

mihot117 reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認させていただきました。
細かい部分となりますが、追加でコメントいたしましたのでご確認いただけますと幸いです。
あともう一息ですね！このまま前進してまいりましょう。

修正後、再度レビュー依頼フォームをご提出ください。

ご対応のほど、よろしくお願いいたします。

sorairo580 committed:
DB設計６

カラム名の‐を修正、purchase_recordsに統一、purchase_recordsテーブルのuserカラムを追加、phone_numberにユニーク制約を追加（ER図に記述しているため）

sorairo580 commented:
度々申し訳ありません。
DB設計６のプッシュを行いました。
ご確認をお願いします。

ER図↓
https://gyazo.com/ccfd95577a2779e6b406109a0f7cd72c

yuri240 commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| room_number   | string     |                                |
+| phone_number  | string     | null: false, unique: true      |
+| item          | references | null: false, foreign_key: true |
+| user          | references | null: false, foreign_key: true |
```
itemとuserの外部キーカラムは不要です。
代わりにpurchase_recordカラムを用意しましょう！

理由：
購入が行われるのと同時に配送先住所の登録が完了するためです。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| city          | string     | null: false                    |
+| house_number  | string     | null: false                    |
+| room_number   | string     |                                |
+| phone_number  | string     | null: false, unique: true      |
```
`| phone_number  | string     | null: false, unique: true      |`

電話番号に一意性制約は不要となります。
Optionsを修正しましょう！

yuri240 commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| prifecture_id | integer    | null: false                    |
+| city          | string     | null: false                    |
+| house_number  | string     | null: false                    |
+| room_number   | string     |                                |
```
`| room_number   | string     |                                |`

上記は建物名のカラムとなりますので、
`building_name`や`building`がおすすめです。

理由：room_numberですと部屋番号を連想するため。

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
キャプチャ画像と修正点確認させていただきました。

追加でコメントさせていただきましたので、ご確認お願いします。

あと一息かと思いますので頑張りましょう！

sorairo580 committed:
DB設計７

parchese_recordカラムを作成

gem rubocopの追加と編集
turbolinksの無効化
.gitignoreファイルの編集（.Ds_Storeの追記）

sorairo580 commented:
度々申し訳ありません。
「parchese_recordカラムを作成」し、DB設計７のプッシュを行いました。
また、下記の通りカリキュラムを見直していて気づいた点の修正も行っています。

・gem rubocopの追加と編集
・turbolinksの無効化
・.gitignoreファイルの編集（.Ds_Storeの追記）

ご確認をお願いします。

ER図↓
https://gyazo.com/83a99a7d5c374168bf171c5d94f2dc03

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
お送り頂いた内容を確認致しました。

今回【LGTM】とさせて頂きます。
実装お疲れ様でした！


※一点以下のみ修正の上、マージを行いましょう！

prefectureのスペリング
→商品テーブル・配送先住所テーブルの二箇所を修正

```
※現在のブランチはマージ後、ローカルのmasterブランチに忘れずにプルしましょう！
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
```

引き続き実装頑張っていきましょう！
応援しております！

sorairo580 committed:
prefectureのスペル修正

sorairo580 commented:
ありがとうございました！
